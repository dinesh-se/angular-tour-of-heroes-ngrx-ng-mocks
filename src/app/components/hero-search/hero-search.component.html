<div id="search-component">
  <label for="search-box">Hero Search</label>
  <ng-container *ngIf="{ query: searchQuery$ | async } as search">
    <input #searchBox id="search-box" value="{{search.query}}"/>
  </ng-container>

  <ul class="search-result">
    <li *ngFor="let hero of searchResults$ | async" >
      <a routerLink="/detail/{{hero.id}}">
        {{hero.name}}
      </a>
    </li>
  </ul>
</div>
