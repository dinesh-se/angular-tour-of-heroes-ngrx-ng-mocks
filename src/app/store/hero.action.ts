import { createAction, props } from '@ngrx/store';

import { Hero } from '@model/hero';

export const loadHeroes = createAction('[Heroes] Load Heroes');

export const loadHeroesSuccess = createAction(
  '[Heroes] Load Heroes Success',
  props<{ heroes: Hero[] }>()
);

export const loadHeroesFailure = createAction(
  '[Heroes] Load Heroes Failure',
  props<{ error: any }>()
);

export const searchHeroes = createAction(
  '[Heroes] Search Heroes',
  props<{ query: string }>()
);

export const searchHeroesSuccess = createAction(
  '[Heroes] Search Heroes Success',
  props<{ searchResults: Hero[] }>()
);

export const searchHeroesFailure = createAction(
  '[Heroes] Search Heroes Failure',
  props<{ error: any }>()
);

export const addHero = createAction(
  '[Heroes] Add Hero',
  props<{ name: string }>()
);

export const addHeroSuccess = createAction(
  '[Heroes] Add Hero Success',
  props<{ hero: Hero }>()
);

export const addHeroFailure = createAction(
  '[Heroes] Add Hero Failure',
  props<{ error: any }>()
);

export const updateHero = createAction(
  '[Heroes] Update Hero',
  props<{ hero: Hero }>()
);

export const updateHeroSuccess = createAction(
  '[Heroes] Update Hero Success',
  props<{ hero: Hero }>()
);

export const updateHeroFailure = createAction(
  '[Heroes] Update Hero Failure',
  props<{ error: any }>()
);

export const deleteHero = createAction(
  '[Heroes] Delete Hero',
  props<{ id: number }>()
);

export const deleteHeroSuccess = createAction(
  '[Heroes] Delete Hero Success',
  props<{ id: number }>()
);

export const deleteHeroFailure = createAction(
  '[Heroes] Delete Hero Failure',
  props<{ error: any }>()
);

export const HeroActions = {
  loadHeroes,
  loadHeroesSuccess,
  loadHeroesFailure,
  searchHeroes,
  searchHeroesSuccess,
  searchHeroesFailure,
  addHero,
  addHeroSuccess,
  addHeroFailure,
  updateHero,
  updateHeroSuccess,
  updateHeroFailure,
  deleteHero,
  deleteHeroSuccess,
  deleteHeroFailure
};
