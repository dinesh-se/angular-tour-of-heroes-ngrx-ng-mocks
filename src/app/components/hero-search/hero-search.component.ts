import { AsyncPipe, NgFor, NgIf } from '@angular/common';
import { AfterViewInit, Component, ElementRef, OnDestroy, OnInit, ViewChild } from '@angular/core';
import { RouterLink } from '@angular/router';
import { Observable, Subscription, fromEvent } from 'rxjs';
import {
  debounceTime,
  distinctUntilChanged,
  map
} from 'rxjs/operators';

import { Hero } from '@model/hero';
import { Store } from '@ngrx/store';
import { HeroesState } from '@store/hero.reducer';
import { HeroActions } from '@store/hero.action';
import { HeroSelectors } from '@store/hero.selector';

@Component({
  selector: 'app-hero-search',
  standalone: true,
  imports: [NgFor, RouterLink, AsyncPipe, NgIf],
  templateUrl: './hero-search.component.html',
  styleUrl: './hero-search.component.scss'
})
export class HeroSearchComponent implements OnInit, AfterViewInit, OnDestroy {
  searchResults$!: Observable<Hero[]>;
  searchQuery$!: Observable<string>;

  @ViewChild('searchBox') searchBox!: ElementRef;

  private inputListener = new Subscription();

  constructor(private store: Store<HeroesState>) {}

  ngOnInit(): void {
    this.searchResults$ = this.store.select(HeroSelectors.selectSearchResults);
    this.searchQuery$ = this.store.select(HeroSelectors.selectQuery);
  }

  ngAfterViewInit(): void {
    this.inputListener = fromEvent(this.searchBox.nativeElement, 'input').pipe(
      map((event: any) => event.target.value),
      debounceTime(300),
      distinctUntilChanged()
    ).subscribe((query) => {
      console.log('CAME');
      this.store.dispatch(HeroActions.searchHeroes({ query }));
    });
  }

  ngOnDestroy(): void {
    this.inputListener.unsubscribe();
  }
}
