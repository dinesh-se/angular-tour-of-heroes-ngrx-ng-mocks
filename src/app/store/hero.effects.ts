import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { Observable, of } from 'rxjs';
import { catchError, map, switchMap, tap } from 'rxjs/operators';
import { Action } from '@ngrx/store';

import { HeroService } from '@services/hero.service';
import * as HeroActions from './hero.action';

@Injectable()
export class HeroEffects {
  constructor(
    private actions$: Actions,
    private heroService: HeroService
  ) {}

  loadHeroes$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.loadHeroes),
      switchMap(() => this.heroService.getHeroes().pipe(
        tap((heroes) => console.log('HEROES', heroes)),
        map((heroes) => HeroActions.loadHeroesSuccess({ heroes })),
        catchError((error) => of(HeroActions.loadHeroesFailure({ error })))
      ))
    )
  );

  searchHeroes$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.searchHeroes),
      switchMap(({ query }) =>
        this.heroService.searchHeroes(query).pipe(
          map((searchResults) => HeroActions.searchHeroesSuccess({ searchResults })),
          catchError((error) => of(HeroActions.searchHeroesFailure({ error })))
        )
      )
    )
  );

  addHero$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.addHero),
      switchMap(({ name }) =>
        this.heroService.addHero({ name }).pipe(
          map((addedHero) => HeroActions.addHeroSuccess({ hero: addedHero })),
          catchError((error) => of(HeroActions.addHeroFailure({ error })))
        )
      )
    )
  );

  updateHero$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.updateHero),
      switchMap(({ hero }) =>
        this.heroService.updateHero(hero).pipe(
          map(() => HeroActions.updateHeroSuccess({ hero })),
          catchError((error) => of(HeroActions.updateHeroFailure({ error })))
        )
      )
    )
  );

  deleteHero$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.deleteHero),
      switchMap(({ id }) =>
        this.heroService.deleteHero(id).pipe(
          map(() => HeroActions.deleteHeroSuccess({ id })),
          catchError((error) => of(HeroActions.deleteHeroFailure({ error })))
        )
      )
    )
  );
}
