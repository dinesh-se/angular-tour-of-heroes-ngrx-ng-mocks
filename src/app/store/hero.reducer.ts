import { createReducer, on, Action } from '@ngrx/store';

import * as HeroActions from './hero.action';
import { Hero } from '@model/hero';

// Define the shape of the state
export interface HeroesState {
  searchResults: Hero[];
  heroes: Hero[];
  loading: boolean;
  error: any;
}

// Define the initial state
export const initialState: HeroesState = {
  searchResults: [],
  heroes: [],
  loading: false,
  error: null
};

// Create the reducer function using createReducer
const heroesReducer = createReducer(
  initialState,

  // Load Heroes actions
  on(HeroActions.loadHeroes, state => ({
    ...state,
    loading: true,
    error: null
  })),
  on(HeroActions.loadHeroesSuccess, (state, { heroes }) => ({
    ...state,
    heroes,
    loading: false
  })),
  on(HeroActions.loadHeroesFailure, (state, { error }) => ({
    ...state,
    loading: false,
    error
  })),


  on(HeroActions.searchHeroesSuccess, (state, { searchResults }) => ({
    ...state,
    searchResults,
  })),
  on(HeroActions.searchHeroesFailure, (state, { error }) => ({
    ...state,
    error
  })),

  // Add Hero actions
  on(HeroActions.addHeroSuccess, (state, { hero }) => ({
    ...state,
    heroes: [...state.heroes, hero]
  })),
  on(HeroActions.addHeroFailure, (state, { error }) => ({
    ...state,
    error
  })),

  // Update Hero actions
  on(HeroActions.updateHeroSuccess, (state, { hero }) => ({
    ...state,
    heroes: state.heroes.map(h => (h.id === hero.id ? hero : h))
  })),
  on(HeroActions.updateHeroFailure, (state, { error }) => ({
    ...state,
    error
  })),

  // Delete Hero actions
  on(HeroActions.deleteHeroSuccess, (state, { id }) => ({
    ...state,
    heroes: state.heroes.filter(h => h.id !== id)
  })),
  on(HeroActions.deleteHeroFailure, (state, { error }) => ({
    ...state,
    error
  }))
);

// Export the reducer function
export function reducer(state: HeroesState | undefined, action: Action) {
  return heroesReducer(state, action);
}
