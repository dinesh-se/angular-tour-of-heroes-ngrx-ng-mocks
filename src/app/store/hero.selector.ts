import { createSelector, createFeatureSelector } from '@ngrx/store';
import { HeroesState } from './hero.reducer';
import { Hero } from '@model/hero';

export const selectHeroesState = createFeatureSelector<HeroesState>('heroes');

export const selectAllHeroes = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.heroes
);

export const selectHeroesLoading = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.loading
);

export const selectHeroesError = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.error
);

export const selectSearchResults = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.searchResults
);

export const selectQuery = createSelector(
  selectHeroesState,
  (state: HeroesState) =>  state.query
);

export const selectHeroById = (id: number) => createSelector(
  selectAllHeroes,
  (heroes: Hero[]) => heroes.find(hero => hero.id === id)
);
