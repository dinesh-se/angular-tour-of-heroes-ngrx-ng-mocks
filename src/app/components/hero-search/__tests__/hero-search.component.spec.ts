
import { AsyncPipe, NgFor, NgIf } from '@angular/common';
import { Store } from '@ngrx/store';
import { MockBuilder, MockRender, MockedComponentFixture } from 'ng-mocks';
import { MockStore, createMockStore } from '@ngrx/store/testing';
import { interval, take } from 'rxjs';

import { HeroesState } from '@store/hero.reducer';
import { HeroSelectors } from '@store/hero.selector';
import { HeroActions } from '@store/hero.action';
import { HeroSearchComponent } from '../hero-search.component';

describe('HeroSearchComponent', () => {
  let component: HeroSearchComponent;
  let fixture: MockedComponentFixture<HeroSearchComponent>;
  const store: MockStore<HeroesState> = createMockStore({
    selectors: [
      {
        selector: HeroSelectors.selectSearchResults,
        value: [
          {
            id: 1,
            name: 'Odd Hero 1'
          },
          {
            id: 3,
            name: 'Odd Hero 3'
          },
          {
            id: 5,
            name: 'Odd Hero 5'
          }
        ]
      },
      {
        selector: HeroSelectors.selectQuery,
        value: 'odd'
      }
    ]
  });
  const dispatchSpy = jest.spyOn(store, 'dispatch');
  const searchHeroesSpy = jest.spyOn(HeroActions, 'searchHeroes');

  beforeEach(async () => {
    await MockBuilder(HeroSearchComponent)
      .mock(Store, store)
      .keep(NgFor)
      .keep(NgIf)
      .keep(AsyncPipe);

    fixture = MockRender(HeroSearchComponent);
    component = fixture.point.componentInstance;
    fixture.detectChanges();
  });

  it('should create search with query and results', () => {
    const input = component.searchBox.nativeElement;

    expect(component).toBeTruthy();
    expect(input.value).toBe('odd');
    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch action when keys are pressed', ((done) => {
    const input = component.searchBox.nativeElement;

    input.value = 'new query';
    input.dispatchEvent(new Event('input'));
    // This is a workaround since tick with fakeAsync is not working in a Jest environment in RxJS 7.x
    // Issue: https://github.com/angular/angular/issues/44351
    interval(310).pipe(
      take(1),
    ).subscribe(() => {
      expect(dispatchSpy).toHaveBeenCalled();
      expect(searchHeroesSpy).toHaveBeenCalledWith({ query: 'new query' });
      done();
    });
  }));
});
