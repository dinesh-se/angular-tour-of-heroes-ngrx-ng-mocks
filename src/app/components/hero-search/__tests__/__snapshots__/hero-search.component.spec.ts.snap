// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroSearchComponent should create search with query and results 1`] = `
<mock-render>
  <app-hero-search>
    <div
      id="search-component"
    >
      <label
        for="search-box"
      >
        Hero Search
      </label>
      <input
        id="search-box"
      />
      <ul
        class="search-result"
      >
        <li>
          <a>
             Odd Hero 1 
          </a>
        </li>
        <li>
          <a>
             Odd Hero 3 
          </a>
        </li>
        <li>
          <a>
             Odd Hero 5 
          </a>
        </li>
      </ul>
    </div>
  </app-hero-search>
</mock-render>
`;
